### PR TITLE
add webhook for proventing user editing ADC's ControlPlaneNetwork

### PR DESCRIPTION
**Pull request by gwang550**
**What this PR does / why we need it**:
When user change control plane network in ADC after the cluster has been created, this can cause a change in LB IP resulting the cluster to be disconnected
**Which issue(s) this PR fixes**:

Fixes #

**Describe testing done for PR**:

**Special notes for your reviewer**:

**Release note**:

```release-note

```
**New PR Checklist**

- [x] Ensure PR contains only public links or terms
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Squash the commits in this branch before merge to preserve our git history
- [ ] If this PR is just an idea or POC, use a [Draft PR](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests) instead of a full PR
- [x] Add appropriate [labels](https://github.com/vmware-samples/load-balancer-operator-for-kubernetes/labels) according to what type of issue is being addressed.

### CODE REVIEW

File: api/v1alpha1/akodeploymentconfig_webhook_test.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            oldControlPlane: ControlPlaneNetwork{
+                Name: "VM Network 1",
+                CIDR: "10.1.0.0/24",
+            },
+            newControlPlane: ControlPlaneNetwork{
+                Name: "VM Network 1",
+                CIDR: "10.0.0.0/24",
+            },
+            expectErr:      true,
+        },
```
Comment (HanFa) [inline api/v1alpha1/akodeploymentconfig_webhook_test.go:64]: just nit:  add a test case when the newControlPlane is empty

File: api/v1alpha1/akodeploymentconfig_webhook.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2020 VMware, Inc.
```
Comment (HanFa) [inline api/v1alpha1/akodeploymentconfig_webhook.go:1]: nit: 2022

File: api/v1alpha1/akodeploymentconfig_webhook.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var akodeploymentconfiglog = logf.Log.WithName("akodeploymentconfig-resource")
```
Comment (HanFa) [inline api/v1alpha1/akodeploymentconfig_webhook.go:17]: nit: maybe use camel style? e.g. `akoDeploymentConfigLog`?

File: api/v1alpha1/akodeploymentconfig_webhook_test.go (hunk 1 above)
Comment (gwang550) [reply api/v1alpha1/akodeploymentconfig_webhook_test.go:64]: yeah, it's good idea

File: api/v1alpha1/akodeploymentconfig_webhook.go (hunk 3 above)
Comment (gwang550) [reply api/v1alpha1/akodeploymentconfig_webhook.go:17]: changed

File: api/v1alpha1/akodeploymentconfig_webhook.go (hunk 2 above)
Comment (gwang550) [reply api/v1alpha1/akodeploymentconfig_webhook.go:1]: changed

File: api/v1alpha1/akodeploymentconfig_webhook.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:verbs=create;update;delete,path=/validate-networking-tkg-tanzu-vmware-com-v1alpha1-akodeploymentconfig,mutating=false,failurePolicy=fail,groups=networking.tkg.tanzu.vmware.com,resources=akodeploymentconfigs,versions=v1alpha1,name=vakodeploymentconfig.kb.io, sideEffects=None, admissionReviewVersions=v1;v1alpha1
+
+var _ webhook.Validator = &AKODeploymentConfig{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *AKODeploymentConfig) ValidateCreate() error {
```
Comment (XudongLiuHarold) [inline api/v1alpha1/akodeploymentconfig_webhook.go:30]: Do we need to implement this function even if we don't have to do anything here?

File: api/v1alpha1/akodeploymentconfig_webhook.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var akoDeploymentConfigLog = logf.Log.WithName("akodeploymentconfig-resource")
```
Comment (XudongLiuHarold) [inline api/v1alpha1/akodeploymentconfig_webhook.go:17]: Is it possible to use `go.uber.org/zap` package for logger? Just want to align with other places we use in the repo.

File: api/v1alpha1/akodeploymentconfig_webhook.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		oldADC.Spec.ControlPlaneNetwork.Name != r.Spec.ControlPlaneNetwork.Name) {
+
+		return field.Invalid(field.NewPath("spec", "ControlPlaneNetwork"), r.Spec.ControlPlaneNetwork, "field should not be changed")
+	}
+
+	return nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *AKODeploymentConfig) ValidateDelete() error {
```
Comment (XudongLiuHarold) [inline api/v1alpha1/akodeploymentconfig_webhook.go:55]: Same here, do we must have this function?

File: api/v1alpha1/akodeploymentconfig_webhook.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+func (r *AKODeploymentConfig) ValidateUpdate(old runtime.Object) error {
+	akoDeploymentConfigLog.Info("validate update", "name", r.Name)
+
+	oldADC, ok := old.(*AKODeploymentConfig)
+
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a AKODeploymentConfig but got a %T", old))
+	}
+
+	if oldADC != nil && (oldADC.Spec.ControlPlaneNetwork.CIDR != r.Spec.ControlPlaneNetwork.CIDR ||
```
Comment (XudongLiuHarold) [inline api/v1alpha1/akodeploymentconfig_webhook.go:45]: nit, could you refactor this to make it more readable?
Something like this:
```go
if oldADC != nil {
  if (oldADC.Spec.ControlPlaneNetwork.CIDR != r.Spec.ControlPlaneNetwork.CIDR) || (oldADC.Spec.ControlPlaneNetwork.Name != r.Spec.ControlPlaneNetwork.Name) {
  // throw error
  }
}

**Review by XudongLiuHarold — COMMENTED** [review]
Need some minor changes, but generally looks good, thanks @gwang550 .

We need some tests to verify the webhook works.

File: api/v1alpha1/akodeploymentconfig_webhook.go (hunk 4 above)
Comment (gwang550) [reply api/v1alpha1/akodeploymentconfig_webhook.go:30]: This was generated by kubebuilder, for our case, it's not needed, but I left there as place holder in case for future need.

File: api/v1alpha1/akodeploymentconfig_webhook.go (hunk 5 above)
Comment (gwang550) [reply api/v1alpha1/akodeploymentconfig_webhook.go:17]: This was generated by kubebuilder, I can check to see if we can use the zap package for logger

File: api/v1alpha1/akodeploymentconfig_webhook.go (hunk 6 above)
Comment (gwang550) [reply api/v1alpha1/akodeploymentconfig_webhook.go:55]: Same as above

File: api/v1alpha1/akodeploymentconfig_webhook.go (hunk 7 above)
Comment (gwang550) [reply api/v1alpha1/akodeploymentconfig_webhook.go:45]: sure

File: api/v1alpha1/akodeploymentconfig_webhook.go (hunk 4 above)
Comment (XudongLiuHarold) [reply api/v1alpha1/akodeploymentconfig_webhook.go:30]: O, I see

File: config/webhook/manifests.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
+
+---
```
Comment (HanFa) [inline config/webhook/manifests.yaml:2]: This should be added to the `static.yaml` in the TCE as well. Similar case is Antrea validation webhook: https://github.com/vmware-tanzu/community-edition/blob/9baddd59e122b8513a8d088495f2963a6a34c7b1/addons/packages/antrea/1.2.3/bundle/config/upstream/antrea.yaml#L4634

File: config/crd/patches/webhook_in_akodeploymentconfigs.yaml [hunk 9]
```diff
@@ -3,15 +3,17 @@
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
-  name: akodeploymentconfigs.network.tanzu.vmware.com
+  name: akodeploymentconfigs.networking.tkg.tanzu.vmware.com
 spec:
   conversion:
     strategy: Webhook
```
Comment (HanFa) [inline config/crd/patches/webhook_in_akodeploymentconfigs.yaml:9]: We don't need to do the conversion for ADC so far, since the only ADC version is `v1alpha1`. I recommend to change the strategy to be `None`, like [this example](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/).

File: pkg/haprovider/haprovider.go [hunk 10]
```diff
@@ -56,7 +56,7 @@ func (r *HAProvider) CreateOrUpdateHAService(ctx context.Context, cluster *clust
 		Namespace: cluster.Namespace,
 	}, service); err != nil {
 		if apierrors.IsNotFound(err) {
-			r.log.Info(serviceName + " service doesn't exist, start creating it...")
+			r.log.Info(serviceName + " guangyuan version service doesn't exist, start creating it...")
```
Comment (lubronzhan) [inline pkg/haprovider/haprovider.go:59]: Remember to remove this

File: api/v1alpha1/akodeploymentconfig_webhook.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+	oldADC, ok := old.(*AKODeploymentConfig)
+
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a AKODeploymentConfig but got a %T", old))
+	}
+
+	if oldADC != nil {
+		if (oldADC.Spec.ControlPlaneNetwork.CIDR != r.Spec.ControlPlaneNetwork.CIDR) || (oldADC.Spec.ControlPlaneNetwork.Name != r.Spec.ControlPlaneNetwork.Name) {
+			return field.Invalid(field.NewPath("spec", "ControlPlaneNetwork"), r.Spec.ControlPlaneNetwork, "field should not be changed")
```
Comment (lubronzhan) [inline api/v1alpha1/akodeploymentconfig_webhook.go:48]: Could you show the previous value and the new ControlPlaneNetwork.cidr and ControlPlaneNetwork.name value in the log

**Comment by lubronzhan** [comment]
Do we need another one for disabliing customer labeling AIS for the control plane service? https://github.com/vmware-samples/load-balancer-operator-for-kubernetes/blob/36577ed19f18f35d278c50d319df9e303b787662/pkg/haprovider/haprovider.go#L141-L150

**Comment by gwang550** [comment]
> Do we need another one for disabliing customer labeling AIS for the control plane service?
I think we certainly need another for that case as well, but we can address that in a separate PR now given the webhook function is enabled in this controller.

File: api/v1alpha1/akodeploymentconfig_webhook.go (hunk 11 above)
Comment (gwang550) [reply api/v1alpha1/akodeploymentconfig_webhook.go:48]: This is not shown in the log rather shown instantly when user are editing the prohibited field in ADC. It looks something like following:
`error: akodeploymentconfigs.networking.tkg.tanzu.vmware.com "install-ako-for-management-cluster" could not be patched: admission webhook "vakodeploymentconfig.kb.io" denied the request: spec.ControlPlaneNetwork: Invalid value: v1alpha1.ControlPlaneNetwork{Name:"VM Network 2", CIDR:"10.1.0.0/24"}: field should not be changed
`

File: pkg/haprovider/haprovider.go (hunk 10 above)
Comment (gwang550) [reply pkg/haprovider/haprovider.go:59]: sure, will do

File: config/ytt/ako-operator.yaml [hunk 12]
```diff
@@ -42,13 +42,26 @@ spec:
             - --metrics-addr=127.0.0.1:8080
           command:
             - /manager
-          image: #@ dockerImage()
+          image: gwang550/akoo:test
```
Comment (XudongLiuHarold) [inline config/ytt/ako-operator.yaml:45]: Change this back after the test

**Review by XudongLiuHarold — COMMENTED** [review]
Generally looks good, just one more question, is everything under `api/v1alpha1` generated by `kubebuilder` ?

File: config/ytt/ako-operator.yaml (hunk 12 above)
Comment (gwang550) [reply config/ytt/ako-operator.yaml:45]: Thanks for catching that. The believe the [api/v1alpha1/zz_generated.deepcopy.go](https://github.com/vmware-samples/load-balancer-operator-for-kubernetes/pull/66/files#diff-d900babf741981553f1ae363bc2d0062aab6a95256dfa091b688df589e9a5c5d) was formatted by IDE. The only file generated is webhook file. The test was added by me later.

**Review by XudongLiuHarold — APPROVED** [review]
test if my approval still work

**Review by XudongLiuHarold — APPROVED** [review]
LGTM
